### DISCUSSION TITLE
streaming inflate

### DISCUSSION THREAD

**Discussion by sigurdle** [101arrowz/fflate#204] [General]
With pako I used to do this to decode PNG files. PNG can use several IDAT chunks for data.

	```
	var inflator = new pako.Inflate();

	for (var i = 0; i < idats.length-1; ++i) {
		inflator.push(idats[i], false);
	}
	inflator.push(idats[i], true);

	if (inflator.err) {
		return inflator.msg;
	}
	var srcData = inflator.result;

```
while there's a Inflate with push in fflate, it misses "result". What would be the equivalent in fflate to the above pako code ?
Thanks
Sigurd

**Comment by 101arrowz** [comment]
There's no `result` in case you don't want to have all the compressed data in memory at once. You need a routine to concatenate all the buffers.
```js
// concat([idats[0], idats[1], ...]) concatenates all of the buffers into one
var concat = function(u8Arrays) {
  var finalLen = 0;
  for (var i = 0; i < u8Arrays.length; ++i) {
    finalLen += u8Arrays[i].length;
  }

  var result = new Uint8Array(finalLen);

  for (var i = 0, offset = 0; i < u8Arrays.length; ++i) {
    result.set(u8Arrays[i], offset);
    offset += u8Arrays[i].byteLength;
  }

  return result;
}
```

You can use this in one of two ways. If you have all your IDATs up front, use like this:

```js
var srcData = fflate.unzlibSync(concat(idats));
```

If you still need to add the IDATs one at a time (e.g. if you're decompressing as you read them sequentially from a file), use like this:

```js
var chunks = [];
var inflator = new fflate.Unzlib(function(chunk, final) {
  chunks.append(chunk);
});

inflator.push(idat0);
inflator.push(idat1);
// ...

inflator.push(idatEnd, true);

var srcData = concat(chunks);
```
